pwiz_tools/Skyline/Util/BioMassCalc.cs

```diff
@@ -613,9 +614,26 @@ public string StripLabelsFromFormula(string desc)
                     dictAtomCounts.TryGetValue(unlabeled, out var count); // Get current count of unlabeled version, if any
                     dictAtomCounts[unlabeled] = count + kvp.Value; // Add the heavy version's count to the unlabeled version's count
                     dictAtomCounts.Remove(kvp.Key); // And remove heavy isotope from the formula
+                    // Preserve order - e.g. C3C'4H2O3 comes out as C7H2O3 and not something dependent on dictionary implementation like H2O3C7 etc
+                    var index = atomOrder.IndexOf(kvp.Key);
+                    if (index >= 0)
+                    {
+                        if (atomOrder.Contains(unlabeled))
+                        {
+                            atomOrder.RemoveAt(index); // Formula was mixed heavy and light - e.g. C and C'
+                        }
+                        else
+                        {
+                            atomOrder[index] = unlabeled; // Formula was all heavy - e.g. C' but no C
+                        }
+                    }
                 }
             }
-            return dictAtomCounts.Aggregate(string.Empty, (current, pair) => current + string.Format(CultureInfo.InvariantCulture, @"{0}{1}", pair.Key, (pair.Value>1) ? pair.Value.ToString() : string.Empty)); 
+            return atomOrder.Aggregate(string.Empty, (current, atom) =>

```

Comment:
I think "String.Join" is easier to understand than "Aggregate", so I would do:
```
return string.Join(string.Empty, atomOrder.Select(atom=>atom + (dictAtomCounts[atom] == 1 ? string.Empty : dictAtomCounts[atom].ToString(CultureInfo.InvariantCulture))))
```

Reply:
It seems like Aggregate would be more efficient, though, since it doesn't mess with a separator. I mean, it's there for a reason, right?

Reply:
Hah, no: https://docs.microsoft.com/en-us/dotnet/csharp/how-to/concatenate-multiple-strings ". If optimizing performance is critical, consider the StringBuilder class or the String.Concat or String.Join method to concatenate a collection, instead of Enumerable.Aggregate."

Reply:
I think this is the way:
`            return string.Concat(atomOrder.Select(atom =>
            {
                var atomCount = dictAtomCounts[atom];
                return atomCount > 1 ? $@"{atom}{atomCount.ToString(CultureInfo.InvariantCulture)}" : $@"{atom}";
            })); `

Reply:
or rather
 `            return string.Concat(atomOrder.Select(atom =>
            {
                var atomCount = dictAtomCounts[atom];
                return atomCount > 1 ? $@"{atom}{atomCount.ToString(CultureInfo.InvariantCulture)}" : atom;
            })); `

Reply:
Oooh. Neat. Looks like "String.Join" was added in .Net 4.0.
I am going to find all the places where I did "String.Join(string.Empty,..." and change them to use "String.Concat(".

Reply:
And I should look for the places where I used Aggregate.